### PR TITLE
Replace ===/!== in assert_true/false

### PR DESCRIPTION
**Pull request by gsnedders**
This is relatively common and this does some of the simple cases that I can easily replace with regex.

### CODE REVIEW

**Review by ewilligers — APPROVED** [review]

**Comment by gsnedders** [comment]
I believe the only tests broken in this are:

/content-security-policy/script-src/scripthash-basic-blocked.sub.html
/content-security-policy/script-src/scripthash-ignore-unsafeinline.sub.html
/html/obsolete/requirements-for-implementations/other-elements-attributes-and-apis/document-all.html

**Review by fserb — APPROVED** [review]
for canvas

File: vibration/api-is-present.html
Comment (stephenmcgruer) [inline vibration/api-is-present.html:22]: Optional nit (at the original test :D)

```suggestion
          assert_not_equals(navigator.vibrate, undefined, "navigator.vibrate exists");
```

File: webrtc/RTCPeerConnection-restartIce.https.html
Comment (stephenmcgruer) [inline webrtc/RTCPeerConnection-restartIce.https.html:49]: These appear to be explicitly defined to workaround identifiers that change often. Likely we don't want to make this change.

**Review by stephenmcgruer — APPROVED** [review]
Overall LGTM modulo a few comments.

I am concerned about this causing problems in Chromium's CI due to flaky output for failing asserts. `assert_{true, false}` won't log the value, whilst `assert_{equals, not_equals}` will and so if the value changes for a failing assert, we will get inconsistent results with the baseline. That said, Chromium's poor test infrastructure there is not a reason not to merge this, and we do have a 'solution' for it (marking the entire test failing), so I won't block this.

File: content-security-policy/child-src/child-src-allowed.sub.html
Comment (gsnedders) [reply content-security-policy/child-src/child-src-allowed.sub.html:30]: Goes all the way back to the original Blink import, too!

File: content-security-policy/style-src/style-src-hash-blocked.html
Comment (gsnedders) [reply content-security-policy/style-src/style-src-hash-blocked.html:29]: I'd rather keep this PR relatively minimal. From an API POV, it makes as much sense as it did before.

File: webrtc/RTCPeerConnection-restartIce.https.html
Comment (gsnedders) [reply webrtc/RTCPeerConnection-restartIce.https.html:49]: 😢

File: pointerevents/pointerevent_setpointercapture_relatedtarget.html
Comment (gsnedders) [reply pointerevents/pointerevent_setpointercapture_relatedtarget.html:97]: There's some risk associated with any change of `==`/`!=`, but a quick glance made the vast majority look safe and the lack of regressions suggested that we weren't actually seeking the fuzzy. (There's also risk with `===`/`!==` given we handle +/-0 and NaN differently.)

I'm kinda inclined to make the small change here and deal with things like this separately.

**Comment by gsnedders** [comment]
I've left all the optional work undone to avoid bloating the scope of this, but undone the WebRTC change.

> I am concerned about this causing problems in Chromium's CI due to flaky output for failing asserts. `assert_{true, false}` won't log the value, whilst `assert_{equals, not_equals}` will and so if the value changes for a failing assert, we will get inconsistent results with the baseline. That said, Chromium's poor test infrastructure there is not a reason not to merge this, and we do have a 'solution' for it (marking the entire test failing), so I won't block this.

I think I'd rather land this and see what, if anything, becomes flaky with this. We can then deal with that flake and explicitly document it if it's problematic.

**Review by inexorabletash — APPROVED** [review]
IndexedDB/, cookie-store/, service-workers/, storage/ LGTM

File: IndexedDB/idbcursor_advance_index2.htm
Comment (jan-ivar) [inline IndexedDB/idbcursor_advance_index2.htm:37]: This is replacing `!=`, not `!==` like the commit message states.

`x != null` is a [common pattern](https://twitter.com/getify/status/1274454560610889730) to check valid inputs including `""`, `0` and `false`, but not `null` or `undefined`.

This patch effectively replaces `!=` with `!==`, flipping treatment of all `undefined` values tested.

**Review by jan-ivar — CHANGES_REQUESTED** [review]
This change is not behavior neutral, nor does it match the commit message. I suggest either leaving `!=` and `==` alone, or doing a careful audit.

**Review by jrummell-chromium — APPROVED** [review]
encrypted-media/ LGTM

**Review by joeyparrish — APPROVED** [review]
encrypted-media LGTM

File: IndexedDB/idbcursor_advance_index2.htm
Comment (inexorabletash) [reply IndexedDB/idbcursor_advance_index2.htm:37]: In the case of an IDBRequest's success event for a cursor request (opening a cursor or iterating the cursor), the `result` property is `null` when no more results are present. So "not null" is semantically correct here from the perspective of the Indexed DB spec. 

Agreed, though, that it technically reduces coverage, if an implementation provided undefined here. Of course, the test as written would pass if practically any value was provided e.g. "something" is both not null and will result in a TypeError in the subsequent assertion.

A better assertion would be that `result` is an instance of `IDBCursorWithValue`

Comment (gsnedders) [reply IndexedDB/idbcursor_advance_index2.htm:37]: FWIW: there's a reason why the `==`/`!=` changes were in separate commits; the commit messages should be accurate. Note that, as noted above, `===`/`!==` is also a behavioural change (our asserts are comparable to `Object.is`).

We really don't want to be allowing everything `x != null` in the test suite: I'm unaware of _any_ spec that allows implementations to use `null` _or_ `undefined` as a given return value. We really do want to be as specific as possible when checking return values.

Are there risks here? Yes. That we don't have any regressions in Chrome/Firefox/Safari when running any of these tests is a good sign that tests do mean the specific value, rather than relying on the coercion between types. While I've not looked in depth at the changes, I certainly can't see any case which appears to be relying on type coercion in the comparison.

Comment (jan-ivar) [reply IndexedDB/idbcursor_advance_index2.htm:37]: @gsnedders I agree accuracy in tests is tantamount, which is why I question a semantics-changing regex sweep.

Here's what I recommend. Replace
```js
assert_true(x != null, "writer's intention");
```
with
```js
assert_not_equals(x, null, "writer's intention");
assert_not_equals(x, undefined, "writer's intention");
```
Then leave it to individual domain experts to remove redundancy. This makes it behavior neutral and explicit.

And thanks for doing this! Being able to see failing values in asserts saves so much time. 👍

Comment (jgraham) [reply IndexedDB/idbcursor_advance_index2.htm:37]: FWIW my opinion is this is overkill. If we have made the change and don't see any test result changes across multiple browsers, and don't know of specific cases which we believe are incorrect, the probability of introducing a test bug this way seems acceptably low and in particular is likely no higher than the probability of a test being buggy for many other reasons (e.g. only checked in a single browser to begin with).

I also agree with the point made by @inexorabletash that using `!=[=]` in tests is pretty sketchy from a correctness point of view and we'd almost always be better off making a positive assert on the actual expected value or type. If we're going to spend more effort here let's make that change instead.

**Review by jgraham — APPROVED** [review]

**Review by miketaylr — APPROVED** [review]

File: IndexedDB/idbcursor_advance_index2.htm
Comment (jan-ivar) [reply IndexedDB/idbcursor_advance_index2.htm:37]: @jgraham for `assert_true(x != null)` → `assert_not_equals(x, null)` we wouldn't see test result failures becaise it's more permissive not less.

Comment (jgraham) [reply IndexedDB/idbcursor_advance_index2.htm:37]: But we could see new passes. I'm not claiming there's zero risk here, but I don't think the risk is enough to block the PR.

Comment (jan-ivar) [reply IndexedDB/idbcursor_advance_index2.htm:37]: @jgraham I wouldn't expect new passes, since `expected: FAIL` tests tend to represent missing functionality, so having one accidentally pass from this change could only happen if the test was broken I think. Those tests are also a separate and much smaller sub-population I think.

Instead of looking at this in terms of fail/pass, why not preserve the current accuracy? We know all tests pass under the older stricter requirements, so why not preserve those requirements by adding the extra asserts?

Comment (jan-ivar) [reply IndexedDB/idbcursor_advance_index2.htm:37]: `grep -r "!= null" * | grep assert_true` shows [108 hits](https://gist.github.com/jan-ivar/d2f1cd2b94f6d71f6323e34d57f0206c).

Comment (gsnedders) [reply IndexedDB/idbcursor_advance_index2.htm]: For comparison, `rg 'assert_not_equals\(.+,\s*(null|undefined)\s*[,)]' --no-heading | wc -l` shows 647 hits on master and 791 on this branch, so in many ways this ship has already sailed. I don't totally oppose reverting the changes in this branch, and there's now a couple of fixup commits that get rid of them.

**Review by pwnall — APPROVED** [review]

**Comment by reillyeon** [comment]
This change is too large. I have no idea what part of it @wpt-pr-bot wants me to look at.

**Comment by stephenmcgruer** [comment]
> This change is too large. I have no idea what part of it @wpt-pr-bot wants me to look at.

wpt-pr-bot uses 'reviewers' as a cc list (because GitHub has basically no support for cc... [there is a very stalled RFC](https://github.com/web-platform-tests/rfcs/pull/37) to change how wpt-pr-bot works here), so you are not necessarily expected to review this. GitHub uses 'assignees' for what Chromium calls 'reviewers'.

That said, it looks like you are listed as an owner for 5 directories:

```
$ grep -r reillyeon `find . -name META.yml`
./bluetooth/META.yml:  - reillyeon
./webusb/META.yml:  - reillyeon
./mediacapture-image/META.yml:  - reillyeon
./orientation-event/META.yml:  - reillyeon
./screen-wake-lock/META.yml:  - reillyeon
```

So for this PR, `mediacapture-image/getPhotoCapabilities.html` seems to be the only file that you might want to look at.

(I do not disagree that this is a large and wide PR, and wpt-pr-bot does a bad job for large, wide PRs.)

--------------

Overall, for this PR I am most interested in whether @jan-ivar is happy with the new version (and thanks to @gsnedders for taking on this refactor!)

**Review by Steditor — APPROVED** [review]

**Comment by stephenmcgruer** [comment]
LGTM again. I understand there is some risk of changed semantics here, but with the removal of most of the controversial cases I think this is a net positive change for the test suite.

I took a look at the Chrome, Firefox, and Safari results. Chrome and Firefox had no different results before/after at all. Safari had a few:

`css/css-fonts/variations/at-font-face-font-matching.html` - [results change](https://wpt.fyi/results/css/css-fonts/variations/at-font-face-font-matching.html?diff&filter=ADC&q=is%3Adifferent&run_id=678730007&run_id=674650002) seems suspicious at first glance, but looking at wpt.fyi these tests are pretty flaky on Safari and I accept that this is likely flake rather than a behavioral change.

`resource-timing/resource_dedicated_worker.html` - flaky on Safari

`user-timing/measure_associated_with_navigation_timing.html` - flaky on Safari


Please give 24 hours for someone to raise a new objection, otherwise LGTM to merge.

File: mediacapture-streams/MediaDevices-getUserMedia.https.html
Comment (jan-ivar) [inline mediacapture-streams/MediaDevices-getUserMedia.https.html:18]:
```suggestion
  assert_not_equals(navigator.mediaDevices.getUserMedia, undefined, "navigator.mediaDevices.getUserMedia exists.");
```

File: mediacapture-streams/MediaDevices-getUserMedia.https.html
Comment (jan-ivar) [inline mediacapture-streams/MediaDevices-getUserMedia.https.html:20]:
```suggestion
  assert_not_equals(navigator.mediaDevices.getSupportedConstraints, undefined, "navigator.mediaDevices.getSupportedConstraints exists.");
```

**Review by jan-ivar — APPROVED** [review]

**Pull request merged**
